### PR TITLE
New EDSL, mostly from Biokepi's overlay, fixes #287

### LINKED ISSUE #287: Merge "advanced" API from Biokepi
Though I'd prefer if we called it the "API" and the current API the "base API"; the Biokepi version uses better verbiage and is simpler (although it uses a row type + parametric types, I think it's still easier to understand, and can be understood and used without understanding that).

### PR DESCRIPTION
**Pull request by smondet**
[Review on Reviewable](https://reviewable.io/reviews/hammerlab/ketrew/292)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -199,23 +199,23 @@ The EDSL: Defining Workflows
...
+datastructure to the engine.
 
-A workflow is a graph of targets (nodes).
+A workflow is a graph of “workflow-nodes” (sometimes called “targets”).
 
-There are three kinds of links (edges) between targets:
+There are three kinds of links (edges) between nodes:
 
-- `depends_on`: targets that need to be ensured or satisifed before a target
+- `depends_on`: nodes that need to be ensured or satisifed before a node
```
Comment (arahuja) [inline README.md]: sp: satisfied

File: src/lib/client.mli [hunk 2]
```diff
@@ -163,3 +163,13 @@ val submit:
     One can add tags to all the targets in the workflow before
     submitting with the [add_tags] option.
 *)
+
+val submit:
+  ?override_configuration:Configuration.t ->
+  ?add_tags: string list ->
+  EDSL.user_target ->
+  unit
+(** This is like {!submit_workflow} but for the deprecated/legacy API
```
Comment (arahuja) [inline src/lib/client.mli]: Worth noting what version this is deprecated/legacy in?

File: src/lib/eDSL.ml [hunk 3]
```diff
@@ -208,7 +214,169 @@ let yarn_distributed_shell
...
+        ?equivalence ~tags:(tags @ additional_tags)
+        ~make ()
+      |> (fun x ->
+          if active then Target.activate_exn ~reason:`User x else x)
+  end
+
+type 'a product = 'a
+  constraint 'a = < is_done : Condition.t option ; .. >
+
+(* The main building bloc of the worfklow graph: *)
```
Comment (arahuja) [inline src/lib/eDSL.ml]: bloc

File: src/lib/eDSL.ml [hunk 4]
```diff
@@ -208,7 +214,169 @@ let yarn_distributed_shell
...
+    (product: 'product) : 'product workflow_node =
+  let target_to_submit =
+    let done_when =
+      match done_when with
+      | Some s -> Some s
+      | None -> product#is_done
+    in
+    let depends_on =
+      List.filter_map edges ~f:(function
+        | Depends_on we -> Some we#render
```
Comment (arahuja) [inline src/lib/eDSL.ml]: I'm guessing `we` here is "workflow_edge"? worth spelling out?

**Comment by arahuja** [comment]
I was a little confused on what the new `render`s should be doing? It seems there is one `workflow_node` and `Internal_representation` - maybe I am missing it somewhere, but what should these do?

**Comment by smondet** [comment]
@arahuja I tried to tell users there: https://github.com/hammerlab/ketrew/pull/292/files#diff-fd498ec767e097473c05c520b0e51060R287
and commented more there: https://github.com/hammerlab/ketrew/pull/292/files#diff-fd498ec767e097473c05c520b0e51060R254

It's all for backwards compatibility. I'd have prefered a `#render` function that returns a list of “pure” targets but then some of the code in `Ketrew.Client` would have to be duplicated inside `EDSL`.

File: src/lib/client.mli (hunk 2 above)
Comment (smondet) [inline src/lib/client.mli]: I've put the deprecation version number in `eDSL.mli`.

**Comment by ihodes** [comment]
I have a few comments, this is great though! Thanks for merging in the new API, I think it's a nice improvement.

**Pull request merged**